pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,148 @@
+func (ipAllocator *IPAllocator) ValidateCIDR(cidr string) bool {
+	_, _, err := net.ParseCIDR(cidr)
+	return err == nil
+}
+
+// GetUnreservedIPBlock returns an unreserved /29 IP block.
+// cidr: Provided cidr address in which we need to look for an unreserved /29 IP Block
+// reservedIPRanges: All the Reserved IP ranges present in the cloud provider
+// All the Reserved IP ranges not present in the cloud provider are extracted from the reserved list in the IPAllocator
+// Finally a final reserved list is created by merging these two lists
+// Potential error cases: 1) No /29 Block in the CIDR is unreserved
+// 2) Parsing the CIDR resulted in an error
+func (ipAllocator *IPAllocator) GetUnreservedIPBlock(cidr string, reservedIPRanges map[string]bool) (string, error) {
+	// Merging the reserved lists at the cloud provider end and the service end to form a final reserved list
+	for reservedIPRange := range ipAllocator.reserved {
+		reservedIPRanges[reservedIPRange] = true
+	}
+
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", fmt.Errorf("Unable to parse CIDR %s", cidr)
+	}
+	for cidrIPBlock := cloneIP(ip.Mask(ipnet.Mask)); ipnet.Contains(cidrIPBlock) && err == nil; err = incrementIP(cidrIPBlock, incrementStep29IPBlock) {
+		overLap := false
+		ipRange := fmt.Sprint(cidrIPBlock.String(), reservationBlockSize)
+		for reservedIPRange := range reservedIPRanges {
+			// Find if the current IP range in the CIDR overlaps with any of the reserved IP ranges. If not, this can be returned
+			overLap, err = isOverlap(ipRange, reservedIPRange)
+			// Error while processing cidr is also considered as an overlap to ignore the specific IP block
+			if overLap || err != nil {
+				break
+			}
+		}
+		if !overLap {
+			return ipRange, nil
+		}
+	}
+
+	// No unreserved IP range available in the entire CIDR range. Returning error
+	return "", fmt.Errorf("All of the /29 IP ranges in the cidr %s are reserved", cidr)
+}
+
+// isOverlap checks if two cidrs have any overlapping IPs
+func isOverlap(cidr1 string, cidr2 string) (bool, error) {
+	_, ipnet1, err := net.ParseCIDR(cidr1)
+	// Invalid CIDRs are considered as overlapping
+	if err != nil {
+		return true, fmt.Errorf("Invalid cidr %s provided", cidr1)
+	}
+
+	_, ipnet2, err := net.ParseCIDR(cidr2)
+	if err != nil {
+		return true, fmt.Errorf("Invalid cidr %s provided", cidr2)
+	}
+
+	return ipnet1.Contains(ipnet2.IP) || ipnet2.Contains(ipnet1.IP), nil
+}
+
+// Increment the given IP value by the provided step
+func incrementIP(ip net.IP, step byte) error {
+	ip = ip.To4()
+	for ipByte := 3; ipByte >= 0; ipByte-- {
+		//	If it the least significant byte, add step to it
+		if ipByte == 3 {
+			ip[ipByte] += step
+			// Addition did not result in a carry. We have the incremented value
+			if ip[ipByte] >= step {
+				break
+			}
+		} else {
+			// Add the carry bit from the previous addition. Value of carry can only be 1 as we are performing byte addition
```

Comment:
will we never need `step` sizes over `255`? Genuine question, I don't know.

Reply:
That's an assumption here. Would be required if you want to return /24 < block instead of the existing /29